rclpy/src/rclpy/names.cpp

```diff
@@ -0,0 +1,293 @@
+  {
+    std::string error_text{"topic name '"};
+    error_text += topic;
+    error_text += "' is invalid";
+    throw py::value_error(append_rcl_error(error_text));
+  } else if (ret == RCL_RET_NODE_INVALID_NAME) {
+    std::string error_text{"node name '"};
+    error_text += node_name;
+    error_text += "' is invalid";
+    throw py::value_error(append_rcl_error(error_text));
+  } else if (ret == RCL_RET_NODE_INVALID_NAMESPACE) {
+    std::string error_text{"node namespace '"};
+    error_text += node_namespace;
+    error_text += "' is invalid";
+    throw py::value_error(append_rcl_error(error_text));
+  } else if (ret != RCL_RET_OK) {
+    throw RCLError("failed to do name expansion");
+  }
+
+  return std::string{expanded_topic.get()};
+}
+
+std::string
+remap_topic_name(py::capsule pynode, const char * topic_name)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  // Get the node options
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  if (nullptr == node_options) {
+    throw RCLError("failed to get node options");
+  }
+
+  const rcl_arguments_t * global_args = nullptr;
+  if (node_options->use_global_arguments) {
+    global_args = &(node->context->global_arguments);
+  }
+
+  char * output_cstr = nullptr;
+  rcl_ret_t ret = rcl_remap_topic_name(
+    &(node_options->arguments),
+    global_args,
+    topic_name,
+    rcl_node_get_name(node),
+    rcl_node_get_namespace(node),
+    node_options->allocator,
+    &output_cstr);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to remap topic name");
+  }
+
+  if (nullptr == output_cstr) {
+    // no remapping found
+    return std::string{topic_name};
+  }
+
+  auto name_deleter = [&](char * name) {
+      node_options->allocator.deallocate(name, node_options->allocator.state);
+    };
+  auto remapped_name = std::unique_ptr<char, decltype(name_deleter)>(
+    output_cstr, name_deleter);
+
+  return std::string{remapped_name.get()};
+}
+
+std::string
+resolve_name(py::capsule pynode, const char * topic_name, bool only_expand, bool is_service)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  if (nullptr == node_options) {
+    throw RCLError("failed to get node options");
+  }
+
+  char * output_cstr = nullptr;
+  rcl_ret_t ret = rcl_node_resolve_name(
+    node,
+    topic_name,
+    node_options->allocator,
+    is_service,
+    only_expand,
+    &output_cstr);
+
+  auto name_deleter = [&](char * name) {
+      node_options->allocator.deallocate(name, node_options->allocator.state);
+    };
+  auto resolved_name = std::unique_ptr<char, decltype(name_deleter)>(
+    output_cstr, name_deleter);
```

Comment:
@sloretz FYI, nit^N: [`RCPPUTILS_SCOPE_EXIT`](https://github.com/ros2/rcpputils/blob/71ceb6b0597eee71f91dfeffe77fca7cc1a9cdb4/include/rcpputils/scope_exit.hpp#L65) may be handy in these cases.

Reply:
Noted, I'll use this on the next PR once the rcpputils dependency is added.